### ISSUE TITLE
fortls hangs starting with commit 0d2bfb5db3d20350b4080f81fabe4ef6e677c38e

### ISSUE THREAD

**Issue by mscfd** [hansec/fortran-language-server#122]
For my code base, the fortls python process from the startup phase (right after opening the first fortran file) seems to hang. I can see that the parallel part (the parsing) is done, and then a single python thread runs foreever, at least a couple of minutes (had to kill it). This starts with commit 0d2bfb5db3d20350b4080f81fabe4ef6e677c38e.
Until commit 74cee75dc116ec92e7f5334fa46547139b475284, startup phase took a few seconds. With 187da5554088b390283c04610ae3f753600fa7d5, startup phase was significantly reduced (possibly due to bugs introduced with this commit?)

Otherwise I have had no time to further investigate this issue unfortunately.

**Comment by hansec** [comment]
Thanks for the report, sorry about that. This is likely due to an incorrect circular reference getting created somewhere. Due to the commit that created the problem it is probably either a circular USE reference or a circular object reference due to renaming in a USE ONLY statement. 

Would it be possible for me to get the problematic source code or is it closed source? I would be happy to take a look if I had access to the code. If that isn't possible, you could try doing debug initializations while printing information around the "link resolution" setup at line [1278 in `langserver.py`](https://github.com/hansec/fortran-language-server/blob/414f988e3b9a746ecdf4902a86e1cd628e1a2ab2/fortls/langserver.py#L1278). If you can narrow down the problematic file then hopefully we can sort out the issue.

**Comment by mscfd** [comment]
I cannot share the source code. But I will try to narrow it down and let you know, if I find anything useful.

**Comment by hansec** [comment]
Does this code use submodules?

**Comment by mscfd** [comment]
We use submodules in one instance (turned out that gfortran but also ifort have a number of show stopper bugs...), but these are not concerned here. I have followed fortls and found out, what roughly happens: fortls (with latest patches applied) hangs in objects.py/get_use_tree.
I have added an additional parameter level, to see how far it descends and also to break recursive descend at presribed levels. At about level 6-10 I can see that runtime to complete the first problematic object runtime explodes (exponentially?) until it does not finish in a reasonable time.
I have not investigated the call tree in detail, but I think the following might happen: Our module graph is a big DAC, and core modules at the deepest levels can be reached through many many paths in this DAC. Recursively calling use_tree might travel all those exponentially many paths?

I have added some code which checks whether use_mod was already in use_dict (you do this check to manage the rename_lists), and then I only call get_use_tree recursively if use_mod was new. This seems(!) to solve the problem, at least fortls initialises fast.

Hope this helps.

**Comment by mscfd** [comment]
I just checked the commit who causes this hanging and started wondering. You already use a 'continue' to avoid calling it recursively if use_mod is already present, but only if it has the same only_list. In fact I have found a module, for which there are
'use mod_name'
and 
'use_mod_name, only: xyz'
instances. And it seems that this causes get_use_tree to descend into module mod_name many times, changing between an empty only_list and list [xyz]. Could that happen?

**Comment by hansec** [comment]
Yes, I do try and truncate traversal of the graph whenever possible by identifying paths that have already been taken. However, as you noted it is a little more complicated than just seeing if the module was already visited. Since multiple paths to the same module can import different items depending on "only" statements.

I work with many projects though that have combinations like this and dozens of modules so unless you have a crazy number of modules I don't think it should ever hang unless there is a bug that causes a circular reference or something like that.

**Comment by hansec** [comment]
Have you tried testing the commit referenced above to see if there is an incorrect circular reference getting created somewhere?

**Comment by hansec** [comment]
I believe I may have found the problem. Can you checkout the commit above and let me know if it improves things?

**Comment by mscfd** [comment]
This seems to solve the problem. Thanks a lot! I have simply cherry-picked your commit and checked with v1.10.
BTW: why not use set() instead of lists to manage the only_list. Then 
```
                        if use_dict[use_mod][0].count(only_name) == 0:
                            use_dict[use_mod][0].append(only_name)
```
would reduce to
```
                        use_dict[use_mod][0].add(only_name)
```
and the bug would not have been possible in the first place. No need to count occurences.

Python also has a namedtuple class, which would make reading the code for people like me a bit easier. For example:
```
from collections import namedtuple
...
   use_dict[use_mod] = namedtuple('only_list', 'rename_map')
...
                        use_dict[use_mod].only_list.add(only_name)
```

**Comment by hansec** [comment]
Great, I will release a patch later today. I do plan to replace the list with a set, but since this was a 3 character typo I just wanted to check the simplest fix first. Using `namedtuple` is also a good suggestion, there are a few places where that would improve clarity a lot as you mentioned.

Thanks again for the report and help with debugging!

**Comment by rudrab** [comment]
@mscfd: Can you kindly show me your `fortls` related configuration? I am in `vim8` and using `coc.nvim`. My vimrc is given in #123 .

**Comment by mscfd** [comment]
@rudrab: I am using emacs. In fact I had a similar issue with slow start up, but quite differently. In emacs, projectile determines the project root using git. However, in my project the git root contains many more files (e.g. the build directories contain copies and preprocessed fortran files). Scanning all those files took a lot of time and gave wrong results (but as I said, your issue sounds quite differently). In the end, I just added my own small option to fortls, as I could not convince emacs and its lsp packages to provide the desired root directory. Otherwise I have a very simple setup.

Another small advice: I used the post_message method from language.py to follow the progression of fortls during initialisation. I can upload the commit which makes the connection object available within object.py (used during initialisation) to my fork tomorrow. It might or might not help you to understand your issue. Emacs has an echo area as well as a message buffer, where LSP communications are shown. This way I can easily follow and time progression of fortls by adding post_message calls.
